File: Sort/HeapSort.js
/*
 * Problem: Heap sort
 * Author: Dung Nguyen Tien (Chris)
*/

function heapify(arr, i) {
  var left = 2 * i;
  var right = 2 * i + 1;
  var imax = i;
  
  if (left < arr.length && arr[left] > arr[imax]) imax = left;
  if (right < arr.length && arr[right] > arr[imax]) imax = right;
  
  if (imax != i) {
    if (imax == left) {
      [arr[i], arr[left]] = [arr[left], arr[i]];
      heapify(arr, left);
    }
    else {
      [arr[i], arr[right]] = [arr[right], arr[i]];
      heapify(arr, right);
    }
  }
}

function buildHeap(arr, index) {
  var n = Math.floor(index / 2);
  
  for (var i = n; i >= 0; i--) {
    heapify(arr, i);
  }
}

function heapSort(arr) {
  for (var i = arr.length - 1; i >= 0; i--) {
    buildHeap(arr, i);
    document.writeln(arr[0]);
    [arr[0], arr[i]] = [arr[i], arr[0]];
  }
}

var arr = [1,5,3,8,9,0];
heapSort(arr);
document.writeln(JSON.stringify(arr));

File: Sort/HeapSort.cs
/*
 * Problem: HeapSort
 * Author: Dung Nguyen Tien (Chris)
*/

public class Sort
{
	public static void Heapify(List<int> arr, int n, int i)
	{
		var left = 2 * i;
		var right = 2 * i + 1;
		var imax = i;

		if (left < n && arr[left] > arr[imax]) imax = left;
		if (right < n && arr[right] > arr[imax]) imax = right;

		if (imax == i) return;
		else if (imax == left)
		{
			Swap(arr, left, i);
			Heapify(arr, n, left);
		}
		else if (imax == right)
		{
			Swap(arr, right, i);
			Heapify(arr, n, right);
		}
	}

	public static void BuildHeap(List<int> arr)
	{
		var n = arr.Count / 2 - 1;

		for (var i = n; i >= 0; i--)
		{
			Heapify(arr, arr.Count, i);
		}
	}

	public static void HeapSort(List<int> arr)
	{
		BuildHeap(arr);

		for (var i = arr.Count - 1; i >= 0; i--)
		{
			Swap(arr, i, 0);
			Heapify(arr, i, 0);
		}
	}

	public static void Swap(List<int> arr, int x, int y)
	{
		if (x == y) return;

		var tmp = arr[x];
		arr[x] = arr[y];
		arr[y] = tmp;
	}
}